demonstrations/tutorial_fdhs.py

```diff
@@ -0,0 +1,455 @@
+    
+    Returns:
+        mtilde (List): remaining elements of :math:`\mathfrak{m}` s.t. :math:`\mathfrak{m} = \tilde{\mathfrak{m}} \oplus \mathfrak{h}`.
+        a (List): Cartan subalgebra
+
+    """
+
+    h = [m[which]] # first candidate
+    mtilde = m.copy()
+
+    for m_i in m:
+        if _commutes_with_all(m_i, h):
+            if m_i not in h:
+                h.append(m_i)
+    
+    for h_i in h:
+        mtilde.remove(h_i)
+    
+    return mtilde, h
+
+mtilde, h = cartan_subalgebra(m)
+len(g), len(k), len(mtilde), len(h)
+
+##############################################################################
+# We now have the Cartan decomposition :math:`\mathfrak{g} = \mathfrak{k} \oplus \tilde{\mathfrak{m}} \oplus \mathfrak{h}``
+# and with that all the necessary ingredients for the KAK decomposition.
+# 
+# Variational KhK
+# ---------------
+#
+# Obtaining the actual decomposition is highly non-trivial and there is no canonical way to go about computing it in terms of linear algebra sub-routines.
+# In [#Kökcü]_, the authors propose to find the extrema of the cost function
+# 
+# .. math:: f(\theta) = \langle K(\theta) v K(\theta)^\dagger, H\rangle
+# 
+# where :math:`\langle \cdot, \cdot \rangle` is some inner product (in our case the trace inner product :math:`\langle A, B \rangle = \text{tr}(A^\dagger B)`).
+# This construction uses the operator :math:`v = \sum_j \pi^j h_j \in \mathfrak{h}`
+# that is such that :math:`e^{i t v}` is dense in :math:`e^{i \mathcal{h}}`.
+# The latter means that for any point in :math:`e^{i \mathcal{h}}` there is a :math:`t` such that :math:`e^{i t v}` reaches it.
+# Let us construct it.
+
+gammas = [np.pi**i % 2 for i in range(1, len(h)+1)]
+
+v = qml.dot(gammas, h)
+v_m = qml.matrix(v, wire_order=range(n_wires))
+v_m = jnp.array(v_m)
+
+
+##############################################################################
+# 
+# This procedure has the advantage that we can use an already decomposed ansatz
+# 
+# .. math:: K(\theta) = \prod_j e^{-i \theta_j k_j}
+# 
+# for the vertical unitary.
+# 
+# Now we just have to define the cost function and find an extremum.
+# In this case we are going to use gradient descent to minimize the cost function to a minimum.
+# We are going to use ``jax`` and ``optax`` and write some boilerplate for the optimization procedure.
+
+def run_opt(
+    value_and_grad,
+    theta,
+    n_epochs=500,
+    lr=0.1,
+):
+    """Boilerplate jax optimization"""
+    value_and_grad = jax.jit(jax.value_and_grad(loss))
+    optimizer = optax.lbfgs(learning_rate=lr, memory_size=100)
+    opt_state = optimizer.init(theta)
+
+    energy = []
+    gradients = []
+    thetas = []
+
+    @jax.jit
+    def step(opt_state, theta):
+        val, grad_circuit = value_and_grad(theta)
+        updates, opt_state = optimizer.update(grad_circuit, opt_state, theta, value=val, grad=grad_circuit, value_fn=loss)
```

Comment:
Horizontal scrollbar
```suggestion
        updates, opt_state = optimizer.update(
            grad_circuit, opt_state, theta, value=val, grad=grad_circuit, value_fn=loss
        )
```